src/NHibernate/Type/CollectionType.cs

```diff
@@ -369,19 +370,36 @@ public virtual object InstantiateResult(object original)
 			return result;
 		}
 
-		public virtual object ReplaceElements(object original, object target, object owner, IDictionary copyCache,
-											  ISessionImplementor session)
+		public virtual object ReplaceElements(object original, object target, object owner, IDictionary copyCache, ISessionImplementor session)
 		{
-			// TODO: does not work for EntityMode.DOM4J yet!
-			object result = target;
-			Clear(result);
+			var elemType = GetElementType(session.Factory);
+			var resultPc = target as IPersistentCollection;
+			var originalPc = original as IPersistentCollection;
+			var clearTargetsDirtyFlag = false;
+			if (resultPc != null)
+			{
+				if (originalPc == null)
+				{
+					if (!resultPc.IsDirty && AreCollectionElementsEqual((IEnumerable)original, (IEnumerable)target))
+					{
+						clearTargetsDirtyFlag = true;
```

Comment:
Ok. Will be pushed soon.
(
IMHO, I'm not really a big fan of (code) comments at all. Red tests are the only "comments" that work I believe... But in this case a red test cannot be written for this so... I sort of agree ;)
)